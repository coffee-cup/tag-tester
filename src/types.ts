export interface MetaTag {
  name?: string;
  property?: string;
  content?: string;
  value?: string;
}

export interface TagResult {
  url: string;
  tags: MetaTag[];
}

export enum FilterType {
  Html,
  OpenGraph,
  Twitter,
}

export interface Settings {
  onlyShowRecommended: boolean;
  syncSimilarTags: boolean;
  filters: FilterType[];
}
