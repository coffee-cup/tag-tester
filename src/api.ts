import { TagResult } from "./types";

export const fetchTags = async (url: string): Promise<TagResult> => {
  const query = `page=${encodeURIComponent(url)}`;

  const res = await fetch(`/api/html?${query}`);
  if (res.status !== 200) {
    const e = await res.json();
    throw new Error(e.message);
  }

  const json = await res.json();
  return json as TagResult;
};
