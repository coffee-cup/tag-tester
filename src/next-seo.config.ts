import { DefaultSeoProps } from "next-seo";

const title = "Tag Tester";
const url = "https://tagtester.dev";
const description =
  "Test your websites html, open graph and twitter metadata tags.";
const image = "https://tagtester.dev/tagtester.png";

const config: DefaultSeoProps = {
  title,
  description,
  openGraph: {
    type: "website",
    url,
    site_name: title,
    images: [
      {
        url: image,
        width: 1024,
        height: 1024,
      },
    ],
  },
  twitter: {
    handle: "@jakerunzer",
    cardType: "summary",
  },
};

export default config;
