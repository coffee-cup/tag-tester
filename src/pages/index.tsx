import * as React from "react";
import styled from "@emotion/styled";
import css from "@styled-system/css";
import Layout from "../components/Layout";
import Input from "../components/Input";

const StyledHome = styled.div(css({}));

const Home = () => (
  <Layout>
    <StyledHome>
      <h1>Open Graph Tester</h1>

      <Input value="This is large" large />
      <Input value="This is the value" />
      <Input placeholder="Enter the value here" />
    </StyledHome>
  </Layout>
);

export default Home;
