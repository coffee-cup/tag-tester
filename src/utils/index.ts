export const validUrl = (url: any): boolean => {
  if (url == null || typeof url !== "string" || Array.isArray(url)) {
    return false;
  }

  var pattern = new RegExp(
    "^(https?:\\/\\/)?" + // protocol
    "((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|" + // domain name
    "((\\d{1,3}\\.){3}\\d{1,3}))|" + // OR ip (v4) address
    "localhost" + // OR localhost
    "(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*" + // port and path
    "(\\?[;&a-z\\d%_.~+=-]*)?" + // query string
      "(\\#[-a-z\\d_]*)?$", // fragment locator;
    "i", // fragment locator
  );
  return !!pattern.test(url);
};
