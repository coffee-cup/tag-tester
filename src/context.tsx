import * as React from "react";
import { validUrl } from "./utils";
import { Results } from "./types";

export interface OGState {
  url: string;
  isUrlError: boolean;
  error: string;
  setUrl: (value: string) => void;
  fetchTags: () => void;
  results: Results | null;
}

const OGContext = React.createContext<OGState>({} as OGState);

export const useOG = (): OGState => {
  const state = React.useContext(OGContext);
  return state;
};

export const OGProvider: React.FC = props => {
  const [url, setUrl] = React.useState("https://alfie.prodo.ai");
  const [error, setError] = React.useState<string | null>(null);
  const [isUrlError, setIsUrlError] = React.useState(false);

  const [results, setResults] = React.useState<Results | null>(null);

  const fetchTags = async () => {
    if (isUrlError || url === "") {
      return;
    }

    const query = `page=${encodeURIComponent(url)}`;
    const json = await fetch(`/api/html?${query}`).then(res => res.json());

    if (json.error != null) {
      setError(json.error);
    } else {
      setResults(json);
      setError(null);
    }
  };

  React.useEffect(() => {
    if (results == null) {
      fetchTags();
    }
  }, []);

  const value: OGState = {
    url,
    isUrlError,
    error,
    setUrl: value => {
      setUrl(value);
      setIsUrlError(value !== "" && !validUrl(value));
    },
    fetchTags,
    results,
  };

  return (
    <OGContext.Provider value={value}>{props.children}</OGContext.Provider>
  );
};
