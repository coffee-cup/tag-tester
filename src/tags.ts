import { MetaTag, FilterType } from "./types";
import { rootUrl } from "./utils";

export const getNameProp = (tag: MetaTag): string =>
  tag.name != null ? "name" : "property";

export const getValueProp = (tag: MetaTag): string =>
  tag.content != null ? "content" : "value";

export const getName = (tag: MetaTag): string => tag.name ?? tag.property;

export const getValue = (tag: MetaTag): string => tag.content ?? tag.value;

export const htmlTags = ["title", "description", "image"];

export const twitterPrefix = "twitter:";
export const ogPrefix = "og:";

export const isTwitterTag = (tag: MetaTag): boolean => {
  const name = getName(tag);
  return name != null && name.startsWith(twitterPrefix);
};

export const isOGTag = (tag: MetaTag): boolean => {
  const name = getName(tag);
  return name != null && name.startsWith(ogPrefix);
};

export const isHTMLTag = (
  tag: MetaTag,
  onlyRecommended: boolean = true,
): boolean => {
  const name = getName(tag);
  const isRecommended =
    tag.created || (onlyRecommended ? htmlTags.includes(name) : true);
  return name != null && !isOGTag(tag) && !isTwitterTag(tag) && isRecommended;
};

const imageRegex = /^.*:?image$/;
export const isImageTag = (tag: MetaTag): boolean => {
  const name = getName(tag);
  return imageRegex.test(name);
};

export const makeTag = (
  name: string,
  content: string | undefined,
  created?: boolean,
): MetaTag => {
  const newTag: MetaTag = {
    [name.startsWith(ogPrefix) ? "property" : "name"]: name,
    [name.startsWith(twitterPrefix) ? "value" : "content"]: content,
  };

  if (created) {
    newTag.created = created;
  }

  return newTag;
};

export const createCustomUrl = (
  url: string,
  customTags: { [key: string]: string | undefined },
): string => {
  const query = {
    url,
    ...customTags,
  };

  const queryString = Object.keys(query)
    .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(query[k])}`)
    .join("&");

  return `${rootUrl}/api/page?${queryString}`;
};

export const getFilteredTags = (
  tags: MetaTag[],
  filters: FilterType[],
  onlyRecommended: boolean,
): MetaTag[] =>
  tags.filter(t => {
    if (isHTMLTag(t, onlyRecommended)) {
      return filters.includes(FilterType.Html);
    } else if (isOGTag(t)) {
      return filters.includes(FilterType.OpenGraph);
    } else if (isTwitterTag(t)) {
      return filters.includes(FilterType.Twitter);
    }

    return false;
  });

export const getBase = (s: string): string =>
  s.replace(/^og:/, "").replace(/^twitter:/, "");

export const editTagFromTags = (
  name: string,
  value: string,
  tags: MetaTag[],
  syncSimilar: boolean = false,
): {
  newTags: MetaTag[];
  edited: { [key: string]: string };
} => {
  const edited: { [key: string]: string } = {};
  const newTags = tags.map(t => {
    if (
      t[getNameProp(t)] === name ||
      (syncSimilar && getBase(name) === getBase(t[getNameProp(t)]))
    ) {
      edited[getName(t)] = value;

      return {
        ...t,
        [getValueProp(t)]: value,
      };
    }

    return t;
  });

  return {
    newTags,
    edited,
  };
};

export const deleteTagFromTags = (
  tag: MetaTag,
  tags: MetaTag[],
): {
  newTags: MetaTag[];
  edited: { [key: string]: string | undefined };
} => {
  const newTags = tags.filter(t => t !== tag);
  const edited = {
    [getName(tag)]: undefined,
  };

  return { newTags, edited };
};

export const createNewTag = (
  name: string,
  content: string,
  tags: MetaTag[],
): {
  newTags: MetaTag[];
  edited: { [key: string]: string };
} => {
  const newTag = makeTag(name, content, true);
  const edited = {
    [name]: content,
  };

  return {
    newTags: [...tags, newTag],
    edited,
  };
};
