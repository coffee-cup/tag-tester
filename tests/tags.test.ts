import {
  createCustomUrl,
  getFilteredTags,
  isTwitterTag,
  getNameProp,
  editAllTags,
  getValueProp,
  isOGTag,
  isHTMLTag,
  ogPrefix,
  twitterPrefix,
  getBase,
} from "../src/tags";
import { MetaTag, FilterType } from "../src/types";

const makeTag = (name: string, content: string): MetaTag => ({
  [name.startsWith(ogPrefix) ? "property" : "name"]: name,
  [name.startsWith(twitterPrefix) ? "value" : "content"]: content,
});

describe("tags", () => {
  it("checks if tag is a twitter tag", () => {
    expect(isTwitterTag(makeTag("twitter:image", ""))).toBe(true);
    expect(isTwitterTag(makeTag("og:image", ""))).toBe(false);
    expect(isTwitterTag(makeTag("image", ""))).toBe(false);
  });

  it("checks if tag is a og tag", () => {
    expect(isOGTag(makeTag("twitter:image", ""))).toBe(false);
    expect(isOGTag(makeTag("og:image", ""))).toBe(true);
    expect(isOGTag(makeTag("image", ""))).toBe(false);
  });

  it("checks if tag is an html tag", () => {
    expect(isHTMLTag(makeTag("twitter:image", ""))).toBe(false);
    expect(isHTMLTag(makeTag("og:image", ""))).toBe(false);
    expect(isHTMLTag(makeTag("image", ""))).toBe(true);
  });

  it("check if html tag is recommended", () => {
    expect(isHTMLTag(makeTag("title", ""))).toBe(true);
    expect(isHTMLTag(makeTag("viewport", ""))).toBe(false);
  });

  it("getting name and value props", () => {
    expect(getNameProp(makeTag("image", ""))).toBe("name");
    expect(getNameProp(makeTag("twitter:image", ""))).toBe("name");
    expect(getNameProp(makeTag("og:image", ""))).toBe("property");

    expect(getValueProp(makeTag("image", ""))).toBe("content");
    expect(getValueProp(makeTag("twitter:image", ""))).toBe("value");
    expect(getValueProp(makeTag("og:image", ""))).toBe("content");
  });

  it("creates a custom url", () => {
    expect(createCustomUrl("https://tagtester.dev", {})).toBe(
      "https://tagtester.dev/api/page?url=https%3A%2F%2Ftagtester.dev",
    );
    expect(
      createCustomUrl("https://tagtester.dev", {
        title: "a test",
        desc: "hello",
      }),
    ).toBe(
      "https://tagtester.dev/api/page?url=https%3A%2F%2Ftagtester.dev&title=a%20test&desc=hello",
    );
  });

  it("gets filtered tags", () => {
    const tags = [
      makeTag("title", "this is title"),
      makeTag("viewport", "this is viewport"),
      makeTag("og:image", "this is image"),
      makeTag("twitter:card", "this is card"),
    ];

    expect(getFilteredTags(tags, [], true)).toEqual([]);
    expect(
      getFilteredTags(
        tags,
        [FilterType.Html, FilterType.OpenGraph, FilterType.Twitter],
        true,
      ),
    ).toEqual([tags[0], tags[2], tags[3]]);
    expect(getFilteredTags(tags, [FilterType.Html], true)).toEqual([tags[0]]);
    expect(getFilteredTags(tags, [FilterType.Html], false)).toEqual([
      tags[0],
      tags[1],
    ]);
  });

  it("gets base of tag name", () => {
    expect(getBase("title")).toBe("title");
    expect(getBase("og:title")).toBe("title");
    expect(getBase("twitter:title")).toBe("title");
  });

  it("edits tags", () => {
    const tags = [
      makeTag("title", "this is title"),
      makeTag("viewport", "this is viewport"),
      makeTag("og:title", "this is image"),
      makeTag("twitter:title", "this is card"),
    ];

    {
      const { newTags, edited } = editAllTags(
        "title",
        "new title",
        tags,
        false,
      );
      expect(newTags).toEqual([
        makeTag("title", "new title"),
        tags[1],
        tags[2],
        tags[3],
      ]);
      expect(edited).toEqual({ title: "new title" });
    }

    {
      const { newTags, edited } = editAllTags("title", "new title", tags, true);
      expect(newTags).toEqual([
        makeTag("title", "new title"),
        tags[1],
        makeTag("og:title", "new title"),
        makeTag("twitter:title", "new title"),
      ]);
      expect(edited).toEqual({
        title: "new title",
        "og:title": "new title",
        "twitter:title": "new title",
      });
    }
  });
});
