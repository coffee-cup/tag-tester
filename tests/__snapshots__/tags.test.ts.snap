// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tags can generate html from tags 1`] = `
"<title>this is title</title>
<meta name=\\"viewport\\" content=\\"this is viewport\\" /> 
<meta property=\\"og:title\\" content=\\"this is image\\" /> 
<meta name=\\"twitter:title\\" value=\\"this is card\\" /> "
`;
